lib/Modem.ts

```diff
@@ -0,0 +1,829 @@
+		} finally {
+			// not used
+		}
+
+		return result;
+	}
+
+	/*
+	 * handling for received data from the modem
+	 */
+
+	private _dataReceived(received: Buffer) {
+		this._receivedData += received.toString();
+		const parts = this._receivedData.split('\r\n');
+		this._receivedData = parts.pop() || '';
+
+		console.log(this._receivedData);
+
+		for (const part of parts) {
+			if (part.trim() !== '') {
+				this.events.emitDataReceived(part);
+			}
+
+			for (const newpart of part.split(/\r\n|\n|\r/)) {
+				if (newpart.trim() === '') {
+					continue;
+				}
+
+				// skip echoes from commands
+				if (newpart.startsWith('AT')) {
+					continue;
+				}
+
+				// new incomming SMS
+				if (newpart.startsWith('+CMTI:')) {
+					const smsID: string | undefined = newpart.substring(7).split(',')[1];
+
+					if (smsID && !isNaN(Number(smsID))) {
+						this.events.emitNewSMS(Number(smsID));
+					}
+
+					continue;
+				}
+
+				// new incomming call
+				if (newpart.startsWith('+CLIP')) {
+					const splitted_newpart = newpart.substring(6).split(',');
+
+					if (splitted_newpart[0] && splitted_newpart[1]) {
+						const foundNumbers = /"(.*?)"/g.exec(splitted_newpart[0]);
+
+						this.events.emitNewIncomingCall({
+							phoneNumber: foundNumbers !== null && foundNumbers[1] ? foundNumbers[1] : 'unknown',
+							scheme: splitted_newpart[1]
+						});
+					}
+
+					continue;
+				}
+
+				if (newpart.toUpperCase().includes('^SMMEMFULL')) {
+					this.checkSimMemory();
+
+					continue;
+				}
+
+				if (newpart.startsWith('+CUSD:')) {
+					const splitted_newpart = newpart.substring(7).split(',');
+					const followCode: string | undefined = splitted_newpart[0];
+					let text;
+					let follow;
+
+					switch (splitted_newpart[0]) {
+						case '0':
+							follow = 'no further action required';
+							break;
+						case '1':
+							follow = 'further action required';
+							break;
+						case '2':
+							follow = 'terminated by network';
+							break;
+						case '4':
+							follow = 'operation not supported';
+							break;
+					}
+
+					if (splitted_newpart.length > 1) {
+						const decodable = /"(.*?)"/g.exec(splitted_newpart[1]);
+
+						if (decodable && decodable.length > 1) {
+							text = pdu.utils.Helper.decode16Bit(decodable[1]);
+						} else {
+							text = splitted_newpart[1];
+						}
+					}
+
+					this.events.emitNewIncomingUSSD({ text, follow, followCode });
+					continue;
+				}
+
+				this._receivedCmdResponse.push(newpart);
+
+				if (newpart.trim() === '>' || newpart.trim().toUpperCase() === 'OK' || newpart.toUpperCase().includes('ERROR')) {
+					this.events.emitCommandReponse(this._receivedCmdResponse);
+					this._receivedCmdResponse = [];
+				}
+			}
+		}
+	}
+
+	/*
+	 * private functions
+	 */
+
+	private async checkPinRequired(prio = false) {
+		const response = await this.executeATCommand('AT+CPIN?', true, prio);
+
+		if (!response.startsWith('+CPIN:') || response.toUpperCase().includes('ERROR')) {
```

Comment:
uppercase only once?

Reply:
(in also other place below)

Reply:
Is hopefully improved in the next update